packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs

```diff
@@ -1,99 +1,117 @@
+use std::collections::HashMap;
+
 use async_trait::async_trait;
-use chrono::{DateTime, Utc};
-use error_stack::{IntoReport, Result, ResultExt};
-use futures::{StreamExt, TryStreamExt};
-use tokio_postgres::{GenericClient, RowStream};
-use type_system::uri::{BaseUri, VersionedUri};
+use error_stack::{bail, IntoReport, Report, Result, ResultExt};
+use futures::TryStreamExt;
+use type_system::{uri::BaseUri, EntityType};
 
 use crate::{
-    knowledge::{EntityId, PersistedEntity},
-    ontology::AccountId,
+    knowledge::PersistedEntity,
     store::{
         crud,
-        postgres::parameter_list,
-        query::{EntityQuery, EntityVersion},
+        postgres::resolve::{EntityRecord, PostgresContext},
+        query::{
+            Expression, ExpressionError, Literal, PathSegment, Resolve, ResolveError, Version,
+            UNIMPLEMENTED_LITERAL_OBJECT,
+        },
         AsClient, PostgresStore, QueryError,
     },
 };
 
-async fn by_id_by_latest_version(
-    client: &(impl GenericClient + Sync),
-    entity_id: EntityId,
-) -> Result<RowStream, QueryError> {
-    client
-        .query_raw(
-            r#"
-            SELECT properties, entity_id, entities.version, ids.base_uri, ids.version, created_by
-            FROM entities
-            INNER JOIN ids ON ids.version_id = entities.entity_type_version_id
-            WHERE entity_id = $1 AND entities.version = (
-                SELECT MAX("version")
-                FROM entities
-                WHERE entity_id = $1
-            );
-            "#,
-            parameter_list([&entity_id]),
-        )
-        .await
-        .into_report()
-        .change_context(QueryError)
+async fn resolve_properties(
+    properties: &HashMap<BaseUri, serde_json::Value>,
+    path: &[PathSegment],
+    context: &(impl PostgresContext + Sync),
+) -> Result<Literal, ResolveError> {
+    match path {
+        [] => todo!("{}", UNIMPLEMENTED_LITERAL_OBJECT),
+        [head_path_segment, tail_path_segments @ ..] => {
+            let uri = BaseUri::new(head_path_segment.identifier.clone())
+                .into_report()
+                .change_context(ResolveError::StoreReadError)?;
```

Comment:
`BaseUri::new` will return `ParseBaseUriError`, however, we need to change the context to `ResolveError` as we are resolving properties here. We are not able to resolve properties *because* we are not able to parse the `BaseUri` 🙂